src/procontroller_snowbots/src/ProController.cpp

```diff
@@ -0,0 +1,318 @@
+                        break;
+                    case BTN_WEST:
+                        X(ev.value);
+                        break;
+                    case BTN_NORTH:
+                        Y(ev.value);
+                        break;
+                    case BTN_TL:
+                        leftBumper(ev.value);
+                        break;
+                    case BTN_TR:
+                        rightBumper(ev.value);
+                        break;
+                    case BTN_SELECT:
+                        select(ev.value);
+                        break;
+                    case BTN_START:
+                        start(ev.value);
+                        break;
+                    case BTN_MODE:
+                        home(ev.value);
+                        break;
+                    case ABS_Z:
+                        leftTrigger(ev.value);
+                        break;
+                    case ABS_RZ:
+                        rightTrigger(ev.value);
+                        break;
+                    case ABS_HAT0X:
+                        arrowsRorL(ev.value);
+                        break;
+                    case ABS_HAT0Y:
+                        arrowsUorD(ev.value);
+                        break;
+                    case BTN_THUMBL:
+                        leftJoystickPress(ev.value);
+                        break;
+                    case BTN_THUMBR:
+                        rightJoystickPress(ev.value);
+                        break;
+                }
+                //publish move command and update oldx, oldz
+                if (state == Mode::wheels) {
+                    //Publish motion, update x and z old using tuple
+                    tie(x_old, z_old) = publishMoveXZ(x, z, x_old, z_old);
+                } else if (state == Mode::arm){
+                    publishArmXZ(x,z,x_old,z_old);
+                }
+            }
+        }
+    } while (rc == 1 || rc == 0 || rc == -EAGAIN);
+}
+
+//Prints out a controller event using ROS_INFO
+void ProController::printControllerDebug(int type,int code,int value){
+    auto codeout = libevdev_event_code_get_name(type, code);
+    auto typeout = libevdev_event_type_get_name(type);
+    ROS_INFO("Event: Type: %s Code: %s Value: %d\n", typeout, codeout, value);
+}
+
+//Prints a status message detailing the current control mode
+void ProController::printState(){
+    if (state == Mode::wheels){
+        ROS_INFO("Current mode: controlling wheels");
+    }
+    else if (state == Mode::arm){
+        ROS_INFO("Current mode: controlling arm");
+    }
+    else {
+        ROS_INFO("There is no current mode, which is a problem");
+    }
+}
+
+//If x and z are new commands, this function uses the global pubmove to publish a
+//movement message and return the new or old xz to update readInputs()
+tuple<double,double> ProController::publishMoveXZ(double x_new, double z_new, double x_old, double z_old){
+    if (x_old != x_new || z_old != z_new) {
```

Comment:
Since these are doubles, directly comparing them with `!=` may lead to strange results due to their imprecise nature. Instead, set a tolerance for when two doubles are practically equal, and then compare the difference of the doubles to that tolerance.

e.g.
```cpp
x_old - x_new > 0.0001
```
where 0.0001 is an arbitrarily chosen tolerance value.

Reply:
something to see: https://stackoverflow.com/questions/17333/what-is-the-most-effective-way-for-float-and-double-comparison